### ISSUE TITLE
Int modulo

### ISSUE THREAD

**Issue by davidhesselbom** [cogneco/ooc-kean#508]
`Int modulo(Int)`, defined in `math.ooc`, behaves oddly.

```ooc
import math

(5 % 3) toString() println() // outputs 2.
(5 % -3) toString() println() // also outputs 2.
(-5 % 3) toString() println() // outputs -2, but should output 1. 
// Too bad, but that's how the operator works in C.
((-5) % 3) toString() println() // also outputs -2.
(-5 % -3) toString() println() // also outputs -2.

(5 modulo(3)) toString() println() // outputs 2, as expected.
(5 modulo(-3)) toString() println() // outputs 1. Is this correct?
(-5 modulo(3)) toString() println() // outputs -2, but it should output 1. 
// Unlike the % operator, modulo() was written explicitly to handle this situation,
// but it still doesn't work correctly.
((-5) modulo(3)) toString() println() // outputs 1, which is what we want.
// So... it's not a bug in the `modulo` function itself.
(-5 modulo(-3)) toString() println() // outputs -1. Is this correct?
((-5) modulo(-3)) toString() println() // outputs 2. Is this correct?
```

We're not going to change the behavior of the `%` operator, but the `modulo` function that we added should work the way we want it to. Currently, it doesn't for `-5 modulo(3)`, but what about the cases where the divisor is negative? Should `modulo` always return a positive number?

**Comment by davidpiuva** [comment]
A % B equals the largest X satisfying the equation A + B * I = X < B.

**Comment by sebastianbaginski** [comment]
```ooc
(-5 % 3) toString() println() // outputs -2, but should output 1. 
// Too bad, but that's how the operator works in C.
(-5 % -3) toString() println() // also outputs -2.
```

The result is not specified by the C standard for negative operands. It depends entirely on the compiler, so I guess it is not safe to assume anything about the result in C. It can be `-2` now, but you can switch from `gcc` to `clang` and the same C code suddenly returns `1`.

**Comment by davidpiuva** [comment]
There seems to be many dialects of modulo being defined for negative divisors that different compilers might use.
https://en.wikipedia.org/wiki/Modulo_operation

**Comment by davidhesselbom** [comment]
Okay, but let's forget about the `%` operator for now, and focus on our `modulo` function. What should its behavior be for negative divisors? Should we always return a non-negative value?

> A % B equals the largest X satisfying the equation A + B * I = X < B.

Do you mean `X < |B|` ?

**Comment by davidpiuva** [comment]
|X| < |B| if we choose euclidean modulo.

**Comment by sebastianbaginski** [comment]
```ooc
use ooc-math
import math

a:= -5
(a modulo(3)) toString() println() 
(-5 modulo(3)) toString() println() 
```
gg rock

**Comment by sebastianbaginski** [comment]
I save you the trouble of compiling this, above code prints
>1
>-2

**Comment by davidpiuva** [comment]
So we have a truncated modulo.

**Comment by sebastianbaginski** [comment]
Yeah, and broken parser.

**Comment by sebastianbaginski** [comment]
> Does (-5) in parentheses make any difference?

Yes, it prints `1` in both cases.

> I think the bug is after the parser since it is pre-calculation of constant expressions.
:+1:

**Comment by sebastianbaginski** [comment]
So, for the sake of completeness:
```ooc
use ooc-math
import math

a:= -5
(a modulo(3)) toString() println()
(-5 modulo(3)) toString() println()
((-5) modulo(3)) toString() println()
```
>1
>-2
>1

**Comment by davidpiuva** [comment]
Microsoft's compilers can also give different output for the same input.

**Comment by davidhesselbom** [comment]
I updated the first post, thanks for the `(-5)` find, @sebastianbaginski.

**Comment by marcusnaslund** [comment]
This is holding back several things. Let's reach a decision?

**Comment by marcusnaslund** [comment]
@simonmika says to do it as in `Modulo` in https://github.com/cogneco/Kean/blob/develop/Kean/Math/Single.cs

**Comment by davidpiuva** [comment]
So we get euclidean modulo by adding a rounded absolute dividend in multiples of the divisor when the dividend is negative.

**Comment by davidhesselbom** [comment]
`get` properties don't take arguments, afaik, so how should we make it a property?

**Comment by marcusnaslund** [comment]
Oh sorry, method. The keywords were *not static*.

**Comment by davidhesselbom** [comment]
Anyway, you gave me a link to the `Float` implementation, and we're discussing the `Int` version here.

**Comment by davidpiuva** [comment]
There should be a faster modulo by negating the dividend, and subtracting the result from the divisor.

**Comment by marcusnaslund** [comment]
Yes, but we should follow that **pattern**. The int version is identical: https://github.com/cogneco/Kean/blob/develop/Kean/Math/Integer.cs

**Comment by davidpiuva** [comment]
Multiplying and dividing is quite expensive.

**Comment by davidhesselbom** [comment]
It wouldn't be the first time we multiply and divide things, @davidpiuva... still, I think 
```
public static int Modulo(int dividend, int divisor)
		{
			if (dividend < 0)
				dividend += Integer.Ceiling(Integer.Absolute(dividend) / (float)divisor) * divisor;
			return dividend % divisor;
		}
```
looks unnecessarily complicated. But if Simon says so, let's do it.

**Comment by simonmika** [comment]
Let us implement the above algorithm from C# including tests. If somebody thinks performance is an issue please prove it. And by proving it should have a measurable effect on the time it takes to process a frame.

**Comment by simonmika** [comment]
And if there is a simpler and easier to understand way of doing the above with the same result please do a PR.

**Comment by simonmika** [comment]
And when doing the implementation make sure we get the same result as the above C# code as the behavior of % might depend on virtually everything.

**Comment by marcusnaslund** [comment]
With the `Int64` modulo method implemented with #547 let's make sure the others follow suit really soon, therefore labeling this with `high`.
